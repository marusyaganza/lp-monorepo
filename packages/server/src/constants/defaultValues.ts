import { Language, Role, SortWordsBy, Tense } from '../generated/graphql';

export const WORDS_PER_PAGE = 5;
export const DEFAULT_LANGUAGE = Language.English;
export const DEFAULT_WORDS_SORT_BY = SortWordsBy.UpdatedAt;
export const DEFAULT_ROLE = Role.Member;
export const DEFAULT_TENSE = Tense.Pind;
export const DEFAULT_TOKEN_TTL = 7;
// to clean db every hour + 10 min just in case
export const DEMO_DB_TTL = 4200;
export const DEFAULT_MONGO_CONNECTION = 'mongodb://localhost:27017';
export const DEFAULT_OPTIONS__SAMPLE_SIZE = 50;
