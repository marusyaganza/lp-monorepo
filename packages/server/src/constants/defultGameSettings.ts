import { Game, Language, SettingsByGame } from '../generated/graphql';

export const DEFAULT_GAMES_SETTINGS: SettingsByGame = {
  [Game.Audio]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Image]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Speaking]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.TypeWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectDef]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Conjugation]: {
    wordsPerGame: 2,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  },
  [Game.Gender]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  }
};
