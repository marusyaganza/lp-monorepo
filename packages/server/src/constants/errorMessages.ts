export const ERROR_MESSAGES = {
  NOT_AUTHENTICATED: 'Login to perform this operation',
  NOT_AUTHORIZED: 'You do not have permission to perform this operation',
  INCORRECT_CREDS: 'email or password is incorrect',
  SIGN_UP_FAILED: 'sign up operation failed',
  SING_UP_UNAVAILABLE:
    'Sign up operation is not available now. Sorry for inconvenience.',
  GAME_NOT_FOUND: 'Game not found',
  GAME_GENERATION_FAILED: 'Game data generation failed'
};
