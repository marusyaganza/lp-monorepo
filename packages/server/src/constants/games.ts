import { Game, GameConfig, Language } from '../generated/graphql';
export const GAMES: GameConfig[] = [
  {
    name: 'WordEcho',
    id: 'audio',
    desc: 'Listen, Type, Master. Sharpen language skills through auditory engagement.',
    type: Game.Audio,
    orderNum: 0,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'LexiSense',
    desc: 'Define & Conquer.Match words with their meanings intuitively.',
    id: 'select_definition',
    type: Game.SelectDef,
    orderNum: 1,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Definify: Match & Learn',
    desc: 'Associate definitions with the right words. A captivating word puzzle adventure!',
    id: 'select_word',
    type: Game.SelectWord,
    orderNum: 2,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'WordRecall: Define & Type',
    desc: 'Type words matching given definitions. Boost vocabulary in engaging rounds!',
    id: 'type_word',
    type: Game.TypeWord,
    orderNum: 3,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'PictureWord Blitz',
    desc: 'Identify objects from pictures and type their names to enhance vocabulary skills!',
    id: 'image',
    type: Game.Image,
    orderNum: 4,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'VoiceWord Challenge',
    desc: 'Speak words from clues and images. Sharpen your vocabulary and speaking skills fast!',
    id: 'speaking',
    type: Game.Speaking,
    orderNum: 5,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Conjugation Quest',
    desc: 'Master verb conjugations across tenses and languages in fast-paced, interactive challenges!',
    id: 'conjugation',
    type: Game.Conjugation,
    orderNum: 6,
    languages: [Language.Spanish]
  },
  {
    name: 'GenderWord Challenge',
    desc: 'Choose nouns and match their gender. Test linguistic skills in thrilling rounds!',
    id: 'gender',
    type: Game.Gender,
    orderNum: 7,
    languages: [Language.Spanish]
  }
];
