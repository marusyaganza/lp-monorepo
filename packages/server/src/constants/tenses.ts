import { Tense } from '../generated/graphql';

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto',
  futr: 'Futuro indicativo',
  cond: 'Condicional',
  psub: 'Presente subjentivo'
  // gppt: 'Gerundio'
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
