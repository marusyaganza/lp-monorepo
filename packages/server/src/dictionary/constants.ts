import { Tense } from 'generated/graphql';

export type TagType = {
  tag: string;
  replacement: string;
};

export type ComplexTagType = {
  opening: string;
  closing: string;
};

export const TAGS: TagType[] = [
  { tag: '{b}', replacement: '<b>' },
  { tag: '{/b}', replacement: '</b>' },
  { tag: '{ldquo}', replacement: '\u201C' },
  { tag: '{rdquo}', replacement: '\u201D' },
  { tag: '{it}', replacement: '<i>' },
  { tag: '{/it}', replacement: '</i>' },
  { tag: '{wi}', replacement: '<i>' },
  { tag: '{/wi}', replacement: '</i>' },
  { tag: '{gloss}=', replacement: '<i>' },
  { tag: '{/gloss}', replacement: '</i>' },
  { tag: '{sc}', replacement: '<i>' },
  { tag: '{/sc}', replacement: '</i>' },
  { tag: '{parahw}', replacement: '</i>' },
  { tag: '{/parahw}', replacement: '</i>' },
  { tag: '{phrase}', replacement: '</i>' },
  { tag: '{/phrase}', replacement: '</i>' },
  { tag: '{qword}', replacement: '</i>' },
  { tag: '{/qword}', replacement: '</i>' },
  { tag: '{itsc}', replacement: '<i>' },
  { tag: '{/itsc}', replacement: '</i>' },
  { tag: '{inf}', replacement: '' },
  { tag: '{/inf}', replacement: '' },
  { tag: '{sup}', replacement: '' },
  { tag: '{/sup}', replacement: '' },
  { tag: '{bc}', replacement: '' },
  { tag: '{dx}', replacement: '' },
  { tag: '{/dx}', replacement: '' },
  { tag: '{dx_def}', replacement: '' },
  { tag: '{/dx_def}', replacement: '' },
  { tag: '{dx_ety}', replacement: '' },
  { tag: '{/dx_ety}', replacement: '' },
  { tag: '{ma}', replacement: '' },
  { tag: '{/ma}', replacement: '' },
  { tag: '{rom}', replacement: '' },
  { tag: '{/rom}', replacement: '' }
];

export const COMPLEX_TAGS: ComplexTagType[] = [
  { opening: '{sx', closing: '}' },
  { opening: '{d_link', closing: '}' },
  { opening: '{a_link', closing: '}' },
  { opening: '{i_link', closing: '}' },
  { opening: '{et_link', closing: '}' },
  { opening: '{mat', closing: '}' },
  { opening: '{dxt', closing: '}' }
];

export const COMPLEX_TAGS_TO_REMOVE: ComplexTagType[] = [
  { opening: '{ds|', closing: '}' }
];

export const HW_TAGS: TagType[] = [
  { tag: '{bit}', replacement: '' },
  { tag: '{/bit}', replacement: '' }
];

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto'
  //   futr: 'Futuro indicativo',
  //   cond: 'Condicional',
  //   psub: 'Presente subjentivo',
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
