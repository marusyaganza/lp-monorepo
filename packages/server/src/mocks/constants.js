export const mockEnv = {
  SALT_NUMBER: 3,
  JWT_SECTET: 'JWT_secret'
};
