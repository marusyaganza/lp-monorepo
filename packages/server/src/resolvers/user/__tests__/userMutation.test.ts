import {
  createTestServer,
  getDataFromGQL,
  getErrorMessageFromGQL
} from '../../../tests/helpers';
import { mutations } from '../../../tests/mocks/gqlMutations';
import { Role } from '../../../generated/graphql';
import { models } from '../../../tests/mocks/mockModels';
import {
  mockValidatePassword,
  mockCreateToken,
  mockHashPassword
} from '../../../tests/mocks/mockAuthFunc';
import { ERROR_MESSAGES } from '../../../constants/errorMessages';

const signUpData = {
  firstName: 'User',
  lastName: 'Test',
  email: 'test@test.com',
  password: 'password'
};

const loginInput = { email: 'member@member.com', password: 'password1' };
const originalEnv = process.env;

describe('user mutations', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.resetAllMocks();
    process.env = originalEnv;
  });
  test('login with correct password', async () => {
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      validatePassword: mockValidatePassword,
      models
    });

    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: { input: loginInput }
    });
    expect(mockValidatePassword).toHaveBeenCalledWith('password1', 'password');
    expect(getDataFromGQL(res)).toMatchSnapshot();
    expect(models.User.findOne).toHaveBeenCalledWith({
      email: 'member@member.com'
    });
    expect(mockCreateToken).toHaveBeenCalledWith({ id: '1', role: 'MEMBER' });
  });

  test('login with correct password if token is not defined', async () => {
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      validatePassword: mockValidatePassword,
      models
    });
    // @ts-expect-error: ts do not hanle this mocking case
    mockCreateToken.mockReturnValueOnce(undefined);
    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: { input: loginInput }
    });
    expect(mockValidatePassword).toHaveBeenCalledWith('password1', 'password');
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.SIGN_UP_FAILED);
    expect(models.User.findOne).toHaveBeenCalledWith({
      email: loginInput.email
    });
    expect(mockCreateToken).toHaveBeenCalledWith({
      id: '1',
      role: Role.Member
    });
  });

  test('login with incorrect password', async () => {
    // @ts-expect-error: ts do not hanle this mocking case
    mockValidatePassword.mockReturnValueOnce(false);
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      validatePassword: mockValidatePassword,
      models
    });
    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: { input: loginInput }
    });
    expect(mockValidatePassword).toHaveBeenCalledWith('password1', 'password');
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.INCORRECT_CREDS);
  });

  test('login with invalid email', async () => {
    // @ts-expect-error: ts do not hanle this mocking case
    models.User.findOne.mockReturnValueOnce(null);
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      validatePassword: mockValidatePassword,
      models
    });
    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: { input: loginInput }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.INCORRECT_CREDS);
  });

  test('signUp', async () => {
    // @ts-expect-error: ts do not hanle this mocking case
    models.User.findOne.mockReturnValueOnce(null);
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      models
    });

    const res = await mutate({
      mutation: mutations.signUpMutation,
      variables: {
        input: {
          ...signUpData
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(mockCreateToken).toHaveBeenCalledWith({
      id: '2',
      role: Role.Member
    });
    expect(models.User.createOne).toHaveBeenCalledWith({
      ...signUpData
    });
    expect(getDataFromGQL(res)).toMatchSnapshot();
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
  });

  test('signUp if user creation returns undefined', async () => {
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      hashPassword: mockHashPassword,
      models: {
        ...models,
        User: {
          createOne: jest.fn()
        }
      }
    });

    const res = await mutate({
      mutation: mutations.signUpMutation,
      variables: {
        input: {
          ...signUpData
        }
      }
    });

    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.SIGN_UP_FAILED);
    expect(mockHashPassword).toHaveBeenCalledTimes(0);
    expect(models.User.createOne).toHaveBeenCalledTimes(0);
  });

  test('signUp if token is not defined', async () => {
    // @ts-expect-error: ts do not hanle this mocking case
    models.User.findOne.mockReturnValueOnce(null);
    // @ts-expect-error: ts do not hanle this mocking case
    mockCreateToken.mockReturnValueOnce(undefined);

    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      hashPassword: mockHashPassword,
      models
    });

    const res = await mutate({
      mutation: mutations.signUpMutation,
      variables: {
        input: {
          ...signUpData
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.SIGN_UP_FAILED);
  });

  test('signUp in production', async () => {
    const { mutate } = createTestServer({
      createToken: mockCreateToken,
      hashPassword: mockHashPassword,
      models
    });

    process.env.NODE_ENV = 'production';

    const res = await mutate({
      mutation: mutations.signUpMutation,
      variables: {
        input: {
          ...signUpData
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      ERROR_MESSAGES.SING_UP_UNAVAILABLE
    );
    expect(models.User.findOne).toHaveBeenCalledTimes(0);
    expect(models.User.createOne).toHaveBeenCalledTimes(0);
    expect(mockHashPassword).toHaveBeenCalledTimes(0);
  });
});
