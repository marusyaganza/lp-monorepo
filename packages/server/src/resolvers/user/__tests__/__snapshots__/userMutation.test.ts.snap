// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user mutations login with correct password 1`] = `
{
  "login": {
    "email": "member@member.com",
    "firstName": "User",
    "lastName": "Test",
    "token": "token_{"id":"1","role":"MEMBER"}",
  },
}
`;

exports[`user mutations signUp 1`] = `
{
  "signUp": {
    "email": "test@test.com",
    "firstName": "User",
    "id": "2",
    "lastName": "Test",
    "token": "token_{"id":"2","role":"MEMBER"}",
  },
}
`;
